# Introduction

In this notebook, we'll be exploring the provided car advertisement dataset to flex our data analytics muscles and apply them to a web application. We'll start by importing `pandas` and `plotly_express` to help work with our data.

In [1]:
import pandas as pd
import plotly_express as px

Let's read our csv file into a DataFrame and take a high-level look at it.

In [2]:
df = pd.read_csv("../vehicles_us.csv")
df.info()
df.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


price    model_year       model  condition     cylinders  \
count    51525.000000  47906.000000       51525      51525  46265.000000   
unique            NaN           NaN         100          6           NaN   
top               NaN           NaN  ford f-150  excellent           NaN   
freq              NaN           NaN        2796      24773           NaN   
mean     12132.464920   2009.750470         NaN        NaN      6.125235   
std      10040.803015      6.282065         NaN        NaN      1.660360   
min          1.000000   1908.000000         NaN        NaN      3.000000   
25%       5000.000000   2006.000000         NaN        NaN      4.000000   
50%       9000.000000   2011.000000         NaN        NaN      6.000000   
75%      16839.000000   2014.000000         NaN        NaN      8.000000   
max     375000.000000   2019.000000         NaN        NaN     12.000000   

         fuel       odometer transmission   type paint_color   is_4wd  \
count   51525   43633.000000        51525  51525       42258  25572.0   
unique      5            NaN            3     13          12      NaN   
top       gas            NaN    automatic    SUV       white      NaN   
freq    47288            NaN        46902  12405       10029      NaN   
mean      NaN  115553.461738          NaN    NaN         NaN      1.0   
std       NaN   65094.611341          NaN    NaN         NaN      0.0   
min       NaN       0.000000          NaN    NaN         NaN      1.0   
25%       NaN   70000.000000          NaN    NaN         NaN      1.0   
50%       NaN  113000.000000          NaN    NaN         NaN      1.0   
75%       NaN  155000.000000          NaN    NaN         NaN      1.0   
max       NaN  990000.000000          NaN    NaN         NaN      1.0   

       date_posted  days_listed  
count        51525  51525.00000  
unique         354          NaN  
top     2019-03-17          NaN  
freq           186          NaN  
mean           NaN     39.55476  
std            NaN     28.20427  
min            NaN      0.00000  
25%            NaN     19.00000  
50%            NaN     33.00000  
75%            NaN     53.00000  
max            NaN    271.00000

We can immediately see that some of the columns have missing values, with fewer than the maximum of 51525 values. Depending on what we want to analyze, we'll need to decide how we want to treat these missing values later on. We can also take a look at a sample of the data:

In [3]:
df.head(10)

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
5  14990      2014.0    chrysler 300  excellent        6.0  gas   57954.0   
6  12990      2015.0    toyota camry  excellent        4.0  gas   79212.0   
7  15990      2013.0     honda pilot  excellent        6.0  gas  109473.0   
8  11500      2012.0     kia sorento  excellent        4.0  gas  104174.0   
9   9200      2008.0     honda pilot  excellent        NaN  gas  147191.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28  
5    automatic   sedan       black     1.0  2018-06-20           15  
6    automatic   sedan       white     NaN  2018-12-27           73  
7    automatic     SUV       black     1.0  2019-01-07           68  
8    automatic     SUV         NaN     1.0  2018-07-16           19  
9    automatic     SUV        blue     1.0  2019-02-15           17

The column `is_4wd` seems to use the number 1.0 in place of a Boolean, and the `model` column seems to combine the manufacturer and make into one column. We'll grab the first word to isolate the manufacturer.

## Fixing Data

Before we make the change, let's make sure that the column behaves how we think by checking the unique values in the column.

In [4]:
df['is_4wd'].unique()

array([ 1., nan])

Assuming that 1.0 means `True` and nan is `False`, we can then replace the values.

In [5]:
df['is_4wd'] = df['is_4wd'] == 1.0
df.head(10)

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
5  14990      2014.0    chrysler 300  excellent        6.0  gas   57954.0   
6  12990      2015.0    toyota camry  excellent        4.0  gas   79212.0   
7  15990      2013.0     honda pilot  excellent        6.0  gas  109473.0   
8  11500      2012.0     kia sorento  excellent        4.0  gas  104174.0   
9   9200      2008.0     honda pilot  excellent        NaN  gas  147191.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN    True  2018-06-23           19  
1    automatic  pickup       white    True  2018-10-19           50  
2    automatic   sedan         red   False  2019-02-07           79  
3    automatic  pickup         NaN   False  2019-03-22            9  
4    automatic   sedan       black   False  2019-04-02           28  
5    automatic   sedan       black    True  2018-06-20           15  
6    automatic   sedan       white   False  2018-12-27           73  
7    automatic     SUV       black    True  2019-01-07           68  
8    automatic     SUV         NaN    True  2018-07-16           19  
9    automatic     SUV        blue    True  2019-02-15           17

Now we can take the first word of the `model` column and enrich the data with a new `manufacturer` column.

In [6]:
df['manufacturer'] = df['model'].apply(lambda x: x.split()[0])
manufacturers = sorted(df['manufacturer'].unique())
df.head()

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  \
0    automatic     SUV         NaN    True  2018-06-23           19   
1    automatic  pickup       white    True  2018-10-19           50   
2    automatic   sedan         red   False  2019-02-07           79   
3    automatic  pickup         NaN   False  2019-03-22            9   
4    automatic   sedan       black   False  2019-04-02           28   

  manufacturer  
0          bmw  
1         ford  
2      hyundai  
3         ford  
4     chrysler

# Exploratory Data Analysis

We can now start examining the dataset to try to find any interesting associations. A couple ideas that come to mind are the relationship between `model_year` and `price`, `condition` vs. `days_listed`, or histograms showing the distribution of `odometer` readings, `model_year`, or `days_listed`.

In [7]:
px.histogram(df['days_listed'])

In [8]:
px.histogram(
    df['price'],
    range_x=[0, 75000],
    nbins=700
    )

Let's also create a scatter plot showing price per manufacturer, so that we can see any major differences between them.

In [19]:
px.scatter(
    df,
    x = 'price',
    y = 'model_year',
    labels = {
        'price':'Price',
        'model_year':'Model Year',
        'manufacturer':'Manufacturer'
    },
    color = 'manufacturer',
    title = 'Vehicle Price vs. Model Year by Manufacturer',
    opacity = 0.66
)